trusted_applet/internal/storage/README.md

```diff
@@ -0,0 +1,155 @@
+# Storage
+
+This directory contains a work-in-progress storage system for use with
+the ArmoredWitness.
+It should be considered experimental and subject to change!
+
+Some details on the requirements and design of the storage system are below.
+
+### Requirements
+
+*   Allow the witness unikernel to persist small amounts of data, think multiple independent records of up to a few MB.
+*   Use the eMMC as storage
+*   Avoid artificially shortening the life of storage hardware (flash)
+*   Persisted state should be resilient to corruption from power failure/reboot during writes
+
+#### Nice-to-haves
+
+*   Be somewhat reusable for other ArmoredWitness use cases we may have.
+    *   This probably means being able to store different types of data in specified locations.
+
+#### Non-requirements
+
+*   While we're ultimately limited by the performance of the storage hardware, it's not a priority to achieve the lowest possible latency or highest possible throughput for writes.
+*   Integration with Go's `os.Open()` style APIs (this _would_ be great, but would require upstream work in TamaGo so is explicitly out of scope for now).
+
+#### Out-of-scope
+
+Some things are explicitly out of scope for this design:
+
+*   Protecting against an attacker modifying the data on the storage in some out-of-band fashion,
+*   Hardware failure resulting in previously readable data becoming unreadable/corrupted,
+*   Supporting easy discovery / enumeration of data on disk, or preventing duplicate data from being written. Higher level code should be responsible for understanding what data should be in which slots.
+
+### Design
+
+A relatively simple storage API which offers a fixed number of "storage slots" to which a representation of state can be written. Slot storage will be allocated a range of the underlying storage, starting at a known byte offset and with a known length. This slot storage is also preconfigured with the number of slots that it should allocate (or alternatively/equivalently, the number of bytes to be reserved per-slot).
+
+Each slot is backed by a fixed size "journal" stored across _N_ eMMC sectors. 
+
+Logically it can be thought of like so:
+
+![image showing logical layout](images/logical_layout.png)
+
+Physically it may look like this on the MMC block device itself (8 sectors per journal is just an example):
```

Comment:
Should this be "9 sectors per journal" if it's referring to the diagram?

Reply:
Good spot!